posthog/management/commands/sync_persons_to_clickhouse.py

```diff
@@ -53,61 +56,112 @@ def run(options, sync: bool = False):  # sync used for unittests
-    logger.info(f"Got ${total_pg} in PG and ${len(ch_persons_to_version)} in CH")
+def process_in_batches(get_pg, get_ch, get_key, sync_fn, log_every_nth: int) -> None:
+    last_clickhouse_id = "00000000-0000-0000-0000-000000000000"  # start from the lowest possible UUID
 
-    for i, person in enumerate(persons):
-        if i % (max(total_pg // 10, 1)) == 0 and i > 0:
-            logger.info(f"Processed {i / total_pg * 100}%")
-        ch_version = ch_persons_to_version.get(person.uuid, None)
-        pg_version = person.version or 0
+    # Get objects from Django/PostgreSQL
+    pg_objects = get_pg()
+
+    for i, pg_obj in enumerate(pg_objects):
+        key = get_key(pg_obj)
+        if i % log_every_nth == 0:
+            logger.info(f"Processed {i} objects")
+
+        # If the current PostgreSQL ID is greater than the last ID we got from ClickHouse, fetch a new chunk from ClickHouse
+        # If we didn't get any more new data, we'll mark the last id as the max possible to avoid useless fetches
+        if key > last_clickhouse_id:
+            ch_obj_to_version = get_ch(key)
+            logger.info(f"Fetched from CH after {key}, got {len(ch_obj_to_version)} objects")
+
+            if ch_obj_to_version:
+                last_clickhouse_id = ch_obj_to_version[-1][0]
+            else:
+                last_clickhouse_id = "ffffffff-ffff-ffff-ffff-ffffffffffff"
+
+        # If the current PostgreSQL ID is not in ClickHouse IDs or the version is smaller then sync it
+        ch_version = ch_obj_to_version.get(key, None)
+        pg_version = pg_obj.version or 0
         if ch_version is None or ch_version < pg_version:
-            logger.info(f"Updating {person.uuid} to version {pg_version}")
-            if live_run:
-                # Update ClickHouse via Kafka message
-                create_person(
-                    team_id=team_id,
-                    version=pg_version,
-                    uuid=str(person.uuid),
-                    properties=person.properties,
-                    is_identified=person.is_identified,
-                    created_at=person.created_at,
-                    sync=sync,
-                )
+            logger.info(f"*** Updating {key} to version {pg_version} ***")
+            sync_fn(pg_obj)
         elif ch_version > pg_version:
             logger.info(
-                f"Clickhouse version ({ch_version}) for '{person.uuid}' is higher than in Postgres ({pg_version}). Ignoring."
+                f"Clickhouse version ({ch_version}) for '{key}' is higher than in Postgres ({pg_version}). Ignoring."
+            )
+
+
+def run_person_sync(
+    team_id: int,
+    live_run: bool,
+    deletes: bool,
+    sync: bool,
+    batch_size_pg: int = 100,
+    batch_size_ch: int = 100,
+    log_every_nth: int = 100,
+):
+    logger.info("Running person table sync")
+
+    def get_pg():
+        return Person.objects.filter(team_id=team_id).order_by("id").iterator(chunk_size=batch_size_pg)
```

Comment:
From my reading, this will still load all persons into memory which will still be an issue for large teams. We'd need to implement our own progressive querying here I think.

Reply:
Also, I would prefer specifically selecting fields with `.values_list` which will reduce the amount of processing Django will be doing, and avoid fetching unnecessary data from pg.